Bluetooth: HCI: Use H:4 encoding for buffers

jhedberg opened the pull request:
This PR changes the encoding of HCI packet types from net_buf metadata into the net_buf payload as a H:4 prefix byte. The `bt_buf_get/set_type()` APIs are deprecated but kept compatible with the change (i.e. they do `net_buf_pull/push_u8()` internally).

The PR also removes the experimental H:4 encoding support from the USB transport for several reasons:
- It's non-standard
- It doesn't have any real host-side counter part implementation (even BlueZ upstream doesn't support it)
- It comes with substantial maintenance overhead
- It complicates the HCI raw API needlessly
- It had no CI coverage

Testing status for the drivers. Please check the box once you've tested your respective driver with this PR:

- [x] `h4.c` @jhedberg /upstream CI 
- [x] `h5.c` @jhedberg /upstream CI
- [x] `hci_ambiq.c` @aaronyegx 
- [ ] `hci_da1469x.c` @andrzej-kaczmarek 
- [ ] `hci_esp32.c` @sylvioalves 
- [ ] `hci_ifx_cyw208xx.c` @npal-cy 
- [ ] `hci_ifx_psoc6_bless.c` @npal-cy 
- [x] `hci_nxp.c` @yeaissa
- [x] `hci_silabs_siwx917.c` @jhedberg 
- [x] `hci_silabs_efr32.c` @jhedberg
- [x] `hci_spi_st.c` @HoZHel 
- [x] `hci_stm32wb0.c` @HoZHel 
- [ ] `hci_stm32wba.c` @asm5878 
- [ ] `ipc.c` @dchat-nordic (probably tested by upstream CI?)
- [ ] `ipm_stm32wb.c` @asm5878 
- [x] `spi.c` @HoZHel 
- [ ] `userchan.c` @jhedberg /@aescolar
- [ ] USB @jfischer-no 
- [x] Native controller: @jhedberg @cvinayak /upstream CI

jhedberg commented:
This implements the first item in #88048, which was originally discussed at the recent Bluetooth WG F2F meeting.

alwa-nordic commented:
> The existing bt_buf_set/get_type functions are kept compatible with the change, except that the get() function can only be called once, since it now pulls the byte from the buffer.

I get that this allows us to be partially backwards compatible, but the name of these functions is misleading now. I think the value of keeping them backwards compatible is worth more than a better name, but there may be a way so the function can be invoked multiple times: Doesn't the H4 type always land in the space reserved by `buf.c`? If so, we can just overwrite that byte if the reserved space is already used. Thoughts?

alwa-nordic commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -169,15 +163,14 @@ void bt_buf_rx_freed_cb_set(bt_buf_rx_freed_cb_t cb);
  *  @return A new buffer.
  */
-struct net_buf *bt_buf_get_tx(enum bt_buf_type type, k_timeout_t timeout,
+struct net_buf *bt_buf_get_tx(uint8_t type, k_timeout_t timeout,
```
We need to maintain some level of type information here. Either we should rename the argument to `h4_type`, or seeing as the enum is packed, we keep using the enum here. I suggest adding `BUILD_ASSERT(sizeof(enum bt_buf_type) == sizeof(uint8_t))` in the latter case.

jhedberg replied:
> We need to maintain some level of type information here. Either we should rename the argument to `h4_type`, or seeing as the enum is packed, we keep using the enum here. I suggest adding `BUILD_ASSERT(sizeof(enum bt_buf_type) == sizeof(uint8_t))` in the latter case.

I think it'd be nice to try to use the enum everywhere, combined with the build assert.

jhedberg replied:
I moved from `uint8_t` to `enum bt_buf_type` for these APIs.

alwa-nordic commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -193,26 +186,27 @@ struct net_buf *bt_buf_get_tx(enum bt_buf_type type, k_timeout_t timeout,
  *  @return The BT_* type to of the buffer
  */
-static inline enum bt_buf_type bt_buf_get_type(struct net_buf *buf)
+static inline uint8_t bt_buf_get_type(struct net_buf *buf)
```
I think we can and should deprecate `bt_buf_get_type` and `bt_buf_set_type`, and explicitly do a net_buf pull/push instead. Can we do it in this PR as well?

We need to keep these function for now in deprecated state to remain compatible with out-of-tree drivers. My understanding is that drivers will remain compatible so long as they only call these functions once per buffer, correct?

jhedberg replied:
> I think we can and should deprecate bt_buf_get_type and bt_buf_set_type, and explicitly do a net_buf pull/push instead. Can we do it in this PR as well?

Sure. I'm ok with that.

> My understanding is that drivers will remain compatible so long as they only call these functions once per buffer, correct?

Correct.

jhedberg replied:
Done

alwa-nordic reviewed (changes requested):
Thanks! Looks good!

I have a minor fix that I nack for:
 - https://github.com/zephyrproject-rtos/zephyr/pull/88710#discussion_r2046809914

The other discussions are not hard nacking.

Thalley commented on doc/releases/migration-guide-4.2.rst:
````diff
@@ -227,6 +227,15 @@ Stepper
 Bluetooth
 *********
 
+* The buffer types passing through the HCI driver interface are now indicated as H:4 encoded prefix
+  bytes as part of the buffer payload itself. The bt_buf_set_type() and bt_buf_get_type() functions
+  have been deprecated, but are still useable, with the exception that the latter can only be
+  called once per buffer.
+
+* As part of the HCI buffer encoding change, since packet type values are no longer independent
+  bits, the callback used by :c:func:`bt_buf_rx_freed_cb_set` has been changed to take a bitmask
+  that's constructed using bitwise or of ``BIT(type)``` values.
````
Should these be under a "Drivers" subsection?

jhedberg replied:
> Should these be under a "Drivers" subsection?

I think I'd like to keep it near other Bluetooth item. IIRC we used to have a "Bluetooth HCI" section in past release notes/migration guides, so that might be a good option here.

jhedberg replied:
Moved now to a "Bluetooth HCI" section

Thalley commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -227,6 +227,15 @@ Stepper
 *********
 
+* The buffer types passing through the HCI driver interface are now indicated as H:4 encoded prefix
+  bytes as part of the buffer payload itself. The bt_buf_set_type() and bt_buf_get_type() functions
```
```suggestion
  bytes as part of the buffer payload itself. The :c:func:`bt_buf_set_type` and :c:func:`bt_buf_get_type` functions
```

jhedberg replied:
Done

Thalley commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -499,7 +465,7 @@ static int h4_send(const struct device *dev, struct net_buf *buf)
 	struct h4_data *h4 = dev->data;
 
-	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+	LOG_DBG("buf %p type %u len %u", buf, buf->data[0], buf->len);
```
It's quite implicit that the buffer type is in `data[0]`; should we have a comment here?

jhedberg replied:
It's a fair concern potentially in other places, but this is the H:4 driver which is explicitly specified to use this format, so it shouldn't really be anything special here IMO.

Thalley commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -199,26 +188,28 @@ struct net_buf *bt_buf_get_tx(enum bt_buf_type type, k_timeout_t timeout,
-	((struct bt_buf_data *)net_buf_user_data(buf))->type = type;
+	__ASSERT_NO_MSG(net_buf_headroom(buf) >= 1);
+	net_buf_push_u8(buf, type);
 }
 
-/** Get the buffer type
+/** Get the buffer type. This pulls the H:4 byte prefix from the payload, which means
+ *  that the call can be done only once per buffer.
  *
  *  @param buf   Bluetooth buffer
  *
  *  @return The BT_* type to of the buffer
  */
-static inline enum bt_buf_type bt_buf_get_type(struct net_buf *buf)
+static inline enum bt_buf_type __deprecated bt_buf_get_type(struct net_buf *buf)
 {
-	return (enum bt_buf_type)((struct bt_buf_data *)net_buf_user_data(buf))
-		->type;
+	return net_buf_pull_u8(buf);
 }
```
It may be OK since we are deprecating these, but I think it's a bit dangerous to treat set/push and get/pull as the same thing. 

Could `bt_buf_get_type` not just return `buf->data[0]`?

Thalley replied:
Ditto for `bt_buf_set_type` perhaps

jhedberg replied:
> Could `bt_buf_get_type` not just return `buf->data[0]`

The idea was to try to keep out-of-tree drivers functional without any needed changes, if possible. I imagined the common case to be for the driver to first call `bt_buf_get_type()` before anything else, and if it does that then `pull_u8()` would "just work" with the new payload format.

Thalley replied:
Yeah, that makes sense. Perhaps we should make it very clear in the documentation of the function that it does a pull? (of course since the `buf` parameter is not a `const` it's somewhat implied that it could change already)

jhedberg replied:
IMO it's pretty clearly documented already that these mess with the payload of the buffer:
```c
/** Set the buffer type. The type is encoded as an H:4 byte prefix as part of
 *  the payload itself.
 *
 *  @param buf   Bluetooth buffer
 *  @param type  The BT_* type to set the buffer to
 */
static inline void __deprecated bt_buf_set_type(struct net_buf *buf, enum bt_buf_type type)
{
	__ASSERT_NO_MSG(net_buf_headroom(buf) >= 1);
	net_buf_push_u8(buf, type);
}

/** Get the buffer type. This pulls the H:4 byte prefix from the payload, which means
 *  that the call can be done only once per buffer.
 *
 *  @param buf   Bluetooth buffer
 *
 *  @return The BT_* type to of the buffer
 */
static inline enum bt_buf_type __deprecated bt_buf_get_type(struct net_buf *buf)
{
	return net_buf_pull_u8(buf);
}
```

Thalley commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -150,8 +150,7 @@ static void rx_isr(void)
 				 */
-				buf = bt_buf_get_tx(BT_BUF_H4, K_NO_WAIT,
-						    &type, sizeof(type));
+				buf = bt_buf_get_tx(type, K_NO_WAIT, NULL, 0);
```
Consider updating the documentation for the public function `bt_buf_get_tx` that `data` can be NULL

jhedberg replied:
> Consider updating the documentation for the public function `bt_buf_get_tx` that `data` can be NULL

I can do that, but it's always been the case, i.e. this PR doesn't change anything in that regard.

Thalley replied:
That's not quite true; some of the code you removed accessed the `data` parameter unconditionally : 

```c
	case BT_BUF_H4:
		if (IS_ENABLED(CONFIG_BT_HCI_RAW_H4) &&
		    raw_mode == BT_HCI_RAW_MODE_H4) {
			uint8_t h4_type = ((uint8_t *)data)[0];
```

So previously, as least for `BT_BUF_H4`, `data` could not be NULL

jhedberg replied:
Right, in the H:4 encoding case (for which the experimental USB feature was the only user) that's indeed true. However, that feature was not being enabled anywhere, so for all other cases you could pass NULL. Anyway, I'll add a note to the documentation.

jhedberg replied:
Added note to the documentation. Btw, this is a purely HCI raw API (something that the `buf.h` documentation also doesn't mention anything about. I think we should get rid of APIs like this in order to simplify the HCI interface even further.

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4114,19 +4115,22 @@ static void rx_queue_put(struct net_buf *buf)
 static int bt_recv_unsafe(struct net_buf *buf)
 {
-	bt_monitor_send(bt_monitor_opcode(buf), buf->data, buf->len);
+	/* Don't pull the type, snice we still need it in the rx queue */
```
Suggest to also add this comment to the `bt_send` function that does the same thing

jhedberg replied:
Sure, I'll add a comment

jhedberg replied:
Done

Thalley commented on subsys/usb/device/class/Kconfig.bt:
```diff
@@ -1,18 +1,9 @@
 	select BT_HCI_RAW
 	help
 	  USB Bluetooth device class support
-
-config USB_DEVICE_BLUETOOTH_VS_H4
-	bool "USB Bluetooth H4 vendor command"
-	depends on USB_DEVICE_BLUETOOTH
-	select BT_HCI_RAW_H4
-	select BT_HCI_RAW_CMD_EXT
-	help
-	  Enables vendor command to switch to H:4 transport using the bulk
-	  endpoint.
```
Doesn't this need to be deprecated first?

jhedberg replied:
> Doesn't this need to be deprecated first?

I definitely considered it, but IMO we'd create unreasonable overhead for us with it. See the bullet points I listed in the PR description.

Thalley replied:
Yeah I noticed that you wrote 

> The PR also removes the experimental H:4 encoding support from the USB transport for several reasons:

But it's not marked as experimental and has existed for > 4 years. Maybe the maintainer of USB (@jfischer-no) has some input here. 

If it is truly experimental, then it's fine to remove IMO (although we should possibly add that to the release notes and/or migration guide?)

jfischer-no replied:
The experiment is still ongoing. But "USB Bluetooth H4 vendor command" is not needed (the experiment only uses BT_HCI_RAW_H4 functionality) and can go away.

aaronyegx commented on drivers/bluetooth/hci/hci_ambiq.c:
Verified sample/bluetooth/peripheral_hr worked normally on apollo4p_blue_kxr_evb and apollo3_evb with the change.

alwa-nordic commented on doc/releases/migration-guide-4.2.rst:
```diff
@@ -233,6 +233,18 @@ Bluetooth Audio
+
+* The buffer types passing through the HCI driver interface are now indicated as H:4 encoded prefix
+  bytes as part of the buffer payload itself. The bt_buf_set_type() and bt_buf_get_type() functions
+  have been deprecated, but are still usable, with the exception that the latter can only be
+  called once per buffer.
```
Both of the functions have the same caveat, not just the latter.

Can we be more precise here about the caveat? I think we have to mention that these function will now silently corrupt data unless the following assumptions hold.

 - `bt_buf_get_type(buf)` has to be called exactly once before `buf` is accessed.
 - `bt_buf_set_type(buf)` has to be called exactly once before it's handed to `bt_hci_recv_t`, with no accesses to `buf` in between.

Also, this is the migration guide, not the change log, so we need to say how to migrate away from the deprecated behavior. I.e. Verifying the above assumptions and substitute calls to `bt_buf_get_type` and `bt_buf_set_type` with `net_buf_pull_u8` and `net_buf_push_u8`.

jhedberg replied:
> bt_buf_set_type(buf) has to be called exactly once before it's handed to bt_hci_recv_t, with no accesses to buf in between.

HCI drivers don't generally use this API at all, rather they use the various buffer allocation APIs (like `bt_buf_get_rx()`) which implicitly add the type prefix to the buffer. If you grep under `drivers/bluetooth` you wont find a single user of it.

jhedberg replied:
> Both of the functions have the same caveat, not just the latter.

Fixed.

jhedberg replied:
@alwa-nordic I did some more thinking about this, and I think `hci_types.h` is a justifiable place for the packet type prefix byte. I initially felt against this, since it's different from the rest of the HCI payload, which the drivers will just proxy unmodified back-and-forth, whereas the type is something that needs to be separately handled if the underlying transport doesn't use this format. However, while I made the claim that `drivers/bluetooth.h` should be the *only* interface that drivers should look at, I failed to consider the fact that most drivers either way have to be aware of the packet type-specific header, in order to determine the packet length, and the type-specific header (event, ACL data, etc) is part of `hci_types.h`. Drivers currently also need to peek at the event definitions to make the "discardable" vs "not discardable" distinction.

So, I'll take another stab at keeping `buf.h` (i.e. `enum bt_buf_type`) backwards compatible. I'm still not quite comfortable with keeping "H4" in the other enumeration name, since "H:4" really is about an entire transport and not just about packet types.  I'd rather just strictly document that the values defined based on those values from the transport spec. We don't even have to document it just in reference to H:4, but can say that "the possible values are are commands (1), ACL (2), SCO (3), events (4) and ISO (5) as specified in the UART and 3-Wire UART transport specs" (and extend the reference list to other transports if/when they show up in future specs with the same mapping from value to packet type). So while the documentation makes these references,  I think I'd still prefer to keep the enum name generic, i.e. `enum bt_hci_pkt_type`, however that could be done as a follow-up step and just use our existing H4 defines from `hci_types.h` for now.

jhedberg replied:
I've pushed these changes now. Doesn't look that bad, however there's some inefficiency introduced with the conversions. In some case we even have to do the conversion twice, such as for some bt_hci_raw users like the hci_uart app: For data coming from the host the app first has to convert to `enum bt_buf_type` to be able to call `bt_buf_get_tx()`, and `bt_buf_get_tx()` has to internally convert back to H:4 to be able to prepend the right byte.

jhedberg replied:
And I've now pushed a rebased & squashed patch set. I had to simplify some parts in the commit split, since it wasn't (trivially) possible to include all changes cleanly to the original commit set (I basically recreated all commits as far as it made sense).

alwa-nordic commented on doc/releases/migration-guide-4.2.rst:
````diff
@@ -233,6 +233,18 @@ Bluetooth Audio
+  have been deprecated, but are still usable, with the exception that the latter can only be
+  called once per buffer.
+
+* As part of the HCI buffer encoding change, since packet type values are no longer independent
+  bits, the callback used by :c:func:`bt_buf_rx_freed_cb_set` has been changed to take a bitmask
+  that's constructed using bitwise or of ``BIT(type)``` values.
````
This looks too much like a change log, not a migration guide. It also only says what the new version of API does, and not what it was like before. I miss having that.

Also, the wrong name is referenced. It should be `bt_buf_rx_freed_cb_t`.

`bt_buf_rx_freed_cb_t` used to take a bit-wise and of `enum bt_buf_type`. Now it takes a bit-wise and of `BIT(enum bt_buf_type)`. This is a breaking API change. To upgrade, users must compare `type_mask` against `BIT(enum bt_buf_type)` instead of just `enum bt_buf_type`.

I think we could avoid this breaking change by leaving `enum bt_buf_type` alone (deprecating it) and instead introduce a new type `enum bt_h4_type`. Does that sound feasible to you?

jhedberg replied:
> I think we could avoid this breaking change by leaving `enum bt_buf_type` alone so long as `bt_buf_rx_freed_cb_t` exists and instead introduce a new type `enum bt_h4_type`.

If we go that route, then I'd go all the way and move this to `include/drivers/bluetooth.h` which is the interface between drivers and the host. It'd then also get the namespace of that header i.e. `bt_hci_*`. The HCI spec talks about "packets", so I'd go with `enum bt_hci_pkt_type` or perhaps the shorter `enum bt_hci_pkt`.

That said, I think there's a good chance that we'll eventually completely get rid of `buf.c` and `buf.h` (or at least the latter, leaving the implementation an HCI driver user-side detail). HCI drivers and users of them should only need `include/drivers/bluetooth.h` whereas the current `buf.h` messes this up a bit (it's basically an additional API that a user of the HCI driver API needs to implement).

jhedberg replied:
After some exploration, I'd like to keep `enum bt_buf_type` like it is now in the PR. The reason is that we'd otherwise create more confusion by having *three* different enums for packet types: H4, BT_BUF and BT_HCI_PKT. We'd also need to consider what to do with the likes of `bt_buf_get_rx()` and `bt_buf_get_tx()`, i.e. should they keep taking `enum bt_buf_type` and convert that to the H:4 value, and if so we'd have weird looking driver code where they allocate buffers using bt_buf_type but then push to the payload a H:4 type. AFAIK we also have one known user of this API, i.e. SDC, so I think this breakage is manageable. Also, this API will eventually completely go away when the driver API user (i.e. the host stack) starts proactively giving the driver buffer that it can read incoming data into (this is at least my current vision for how the HCI API could look like).

jhedberg replied:
> Also, the wrong name is referenced. It should be bt_buf_rx_freed_cb_t.

Isn't that the same as "the callback used by :c:func:`bt_buf_rx_freed_cb_set`"? (what the doc now says)

jhedberg replied:
Resolving this comment thread, since I ended up going back to keeping the old enum as-is, i.e. this part goes away from the migration guide.

yeaissa commented:
> @andrzej-kaczmarek @sylvioalves @npal-cy @yeaissa @HoZHel @asm5878 @dchat-nordic: Could you please take a look at the HCI driver testing checklist in this PR description, and mark each driver as tested once you (or someone else with the respective HW) have tested it with this PR? Thanks!

I checked on NXP platform using hci_nxp.c, no regression noticed. Thanks for the change @jhedberg.
I'm not sure I can make the change on your description. Please go ahead and mark it. Thanks.

jhedberg committed:
Bluetooth: HCI: Use H:4 encoding for buffers

Encode the packet type as a H:4 payload prefix for buffers passing to &
from HCI drivers. The existing bt_buf_set/get_type functions are
deprecated, but kept compatible with the change, except that they can only
be called once, since they modify the buffer payload.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
Bluetooth: drivers: Update to use new H:4 buffer encoding

Update all HCI drivers to use the new H:4 encoding for buffers passing
to/from drivers.

One behavioral change that's done in favor of simplicity, is that where
there's previously been switch statements that could return an error for
unsupported packet types now simply pass any received packet unchanged to
lower layers of the controller (or the HCI transport). Handling this is
now the responsibility of the lower layers, however in practice hitting
such scenarios means that there's a mismatch between configured host and
controller features.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
Bluetooth: tests: Update to use new H:4 buffer encoding

Update the Bluetooth tests to assume the new H:4 encoding for data that's
passing between HCI drivers and their users (normally the host stack).

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
Bluetooth: Remove USB H4 mode support

This non-standard feature never had any proper host side implementation
(e.g. it was never upstreamed to BlueZ), and since it comes with notable
maintenance overhead it's fair to just remove it.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

jhedberg committed:
doc: migration-guide-4.2: Add note about Bluetooth HCI changes

Mention the changes that have happened to the Bluetooth HCI driver
interface regarding packet type encoding.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

Thalley commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -48,15 +50,60 @@ enum bt_buf_type {
-/** @brief This is a base type for bt_buf user data. */
-struct bt_buf_data {
-	uint8_t type;
+/** Direction of HCI packets */
+enum bt_buf_dir {
+	BT_BUF_IN,
+	BT_BUF_OUT,
```
From the point of view of the host? 

We could consider making the directions absolute rather than relative like we have for ISO data paths: 

```c
#define BT_HCI_DATAPATH_DIR_HOST_TO_CTLR 0x00
#define BT_HCI_DATAPATH_DIR_CTLR_TO_HOST 0x01
```

The same argument can be applied for the ACL and ISO IN/OUT, but if we modify one, we should probably modify all, which seems outside the scope of this PR

jhedberg replied:
> From the point of view of the host?

From the point of view of `BT_BUF_*_IN/OUT values when converting from a H:4 type (which don't convey any direction info by themselves).

jhedberg replied:
I added a clarifying sentence to the docstring above the enum definition.

jhedberg replied:
Also worth noting that this entire public header will hopefully disappear soon, so I wouldn't put too much effort into trying to micro-optimize it :)

Thalley reviewed (commented):
@jhedberg have you seen any significant changes in memory usage with this PR?

jhedberg commented:
> @jhedberg have you seen any significant changes in memory usage with this PR?

I haven't done any profiling, but I'd expect to see at least a decrease in code size (if you just look at the diffstat). On aggregate it's a big reduction in code, however it'll be spread out over many platforms (a little for each HCI driver) so any single build wont show the same magnitude as the PR as a whole.

jhedberg committed:
Bluetooth: hci_raw: Fix misleading documentation for bt_send()

The documentation for the HCI raw interface was incorrectly talking about
the need to define buffer pools, when in fact the hci_raw module provides
the necessary API to perform buffer allocation.

Signed-off-by: Johan Hedberg <johan.hedberg@silabs.com>

alwa-nordic commented on samples/bluetooth/hci_uart_async/src/hci_uart_async.c:
```diff
@@ -150,7 +150,6 @@ static void send_hw_error(void)
 
 	struct net_buf *buf = bt_buf_get_rx(BT_BUF_EVT, K_FOREVER);
 
-	net_buf_add_u8(buf, BT_HCI_H4_EVT);
```
This line needs to stay, I think. I extended the test to cover this here: https://github.com/zephyrproject-rtos/zephyr/pull/89112. (Feel free to cherry-pick it if you like.)

jhedberg replied:
> This line needs to stay, I think. I extended the test to cover this here: #89112. (Feel free to cherry-pick it if you like.)

The `bt_buf_get_rx()` function (called just above this) already prepends the appropriate type, so the `add_u8()` would cause a double prefix into the buffer.

alwa-nordic reviewed (changes requested):
I am quite happy with how this looks now. I found a small mistake:

Thalley commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -36,6 +36,8 @@ extern "C" {
 
 /** Possible types of buffers passed around the Bluetooth stack in a form of bitmask. */
 enum bt_buf_type {
+	/** Invalid value, used for error checking */
+	BT_BUF_TYPE_INVALID = 0,
```
This is a interesting value; technically any value not defined below are invalid, right?

Consider to name this `BT_BUF_TYPE_NONE`

Thalley commented on include/zephyr/bluetooth/buf.h:
```diff
@@ -48,15 +50,60 @@ enum bt_buf_type {
-struct bt_buf_data {
-	uint8_t type;
+/** Direction of HCI packets. Only used for mapping H:4 to BT_BUF_* values. */
+enum bt_buf_dir {
+	BT_BUF_IN,
+	BT_BUF_OUT,
```
Consider adding documentation for each value here, if only to be able to reach 100% coverage on https://docs.zephyrproject.org/api-coverage/latest/zephyr/bluetooth/buf.h.gcov.html

Thalley reviewed (approved):
2 minor comments, but LGTM
